modules/message-lane/src/lib.rs

```diff
@@ -90,19 +159,237 @@ decl_module! {
+			if queued_messages_dispatched != 0 {
+				frame_support::debug::trace!(
+					target: "runtime",
+					"Spent {} weight on {} queued messages from on_initialize()",
+					weight_spent_on_queued_dispatch,
+					queued_messages_dispatched,
+				);
+			}
+
+			weight_spent_on_queued_dispatch
+		}
+
 		/// Send message over lane.
 		#[weight = 0] // TODO: update me (https://github.com/paritytech/parity-bridges-common/issues/78)
 		pub fn send_message(
 			origin,
 			lane_id: LaneId,
 			payload: T::Payload,
-		) {
-			let _ = ensure_signed(origin)?;
+			delivery_and_dispatch_fee: T::MessageFee,
+		) -> DispatchResult {
+			let submitter = ensure_signed(origin)?;
+
+			// let's first check if message can be delivered to target chain
+			T::TargetHeaderChain::verify_message(&payload).map_err(|err| {
+				frame_support::debug::trace!(
+					target: "runtime",
+					"Message to lane {:?} is rejected by target chain: {:?}",
+					lane_id,
+					err,
+				);
+
+				Error::<T, I>::MessageRejectedByChainVerifier
+			})?;
+
+			// now let's enforce any additional lane rules
+			T::LaneMessageVerifier::verify_message(
+				&submitter,
+				&delivery_and_dispatch_fee,
+				&lane_id,
+				&payload,
+			).map_err(|err| {
+				frame_support::debug::trace!(
+					target: "runtime",
+					"Message to lane {:?} is rejected by lane verifier: {:?}",
+					lane_id,
+					err,
+				);
+
+				Error::<T, I>::MessageRejectedByLaneVerifier
+			})?;
+
+			// let's withdraw delivery and dispatch fee from submitter
+			T::MessageDeliveryAndDispatchPayment::pay_delivery_and_dispatch_fee(
+				&submitter,
+				&delivery_and_dispatch_fee,
+			).map_err(|err| {
+				frame_support::debug::trace!(
+					target: "runtime",
+					"Message to lane {:?} is rejected because submitter {:?} is unable to pay fee {:?}: {:?}",
+					lane_id,
+					submitter,
+					delivery_and_dispatch_fee,
+					err,
+				);
+
+				Error::<T, I>::FailedToWithdrawMessageFee
+			})?;
+
+			// finally, save message in outbound storage and emit event
 			let mut lane = outbound_lane::<T, I>(lane_id);
-			let nonce = lane.send_message(payload);
+			let nonce = lane.send_message(MessageData {
+				payload,
+				fee: delivery_and_dispatch_fee,
+			});
 			lane.prune_messages(T::MaxMessagesToPruneAtOnce::get());
 
+			frame_support::debug::trace!(
+				target: "runtime",
+				"Accepted message {} to lane {:?}",
+				nonce,
+				lane_id,
+			);
+
 			Self::deposit_event(Event::MessageAccepted(lane_id, nonce));
+
+			Ok(())
+		}
+
+		/// Receive messages proof from bridged chain.
+		#[weight = DELIVERY_BASE_WEIGHT + max_dispatch_weight]
+		pub fn receive_messages_proof(origin, proof: MessagesProofOf<T, I>, max_dispatch_weight: Weight) -> DispatchResult {
+			let relayer = ensure_signed(origin)?;
+
+			// verify messages proof && convert proof into messages
+			let messages = T::SourceHeaderChain::verify_messages_proof(proof).map_err(|err| {
```

Comment:
I imagine this will fail if the same messages has already been delivered right? I think in that case, we should still consider using up the remaining `max_dispatch_weight`, which the relayer is willing to pay, to process messages from the queue. In case we reward them with the fee that was payed upfront by the relayer, who actually queued the messages we have a nice incentive to run redundant relayers in case of lane congestion (messages getting queued) and a mechanism that prevents relayers from abusing the queue (i.e. offloading work to `on_initialize`).

Reply:
> I imagine this will fail if the same messages has already been delivered right?

You're right. I've actually missed the fact that we may deliver duplicate messages => relayer should only pay dispatch fee for actually delivered messages. Will respond to that comment once this is solved.

> I think in that case, we should still consider using up the remaining max_dispatch_weight

It is already done - please see how `ByWeightMessageDispatch` works with passed `max_dispatch_weight`.

>  In case we reward them with the fee that was payed upfront by the relayer, who actually queued the messages we have a nice incentive to run redundant relayers in case of lane congestion (messages getting queued) and a mechanism that prevents relayers from abusing the queue (i.e. offloading work to on_initialize).

As I said in other my comments, we should not reward redundant relayers, or txpool of target chain will be flooded with `deliver_transaction(empty-messages, max-transaction-weight)`.

Reply:
> It is already done - please see how `ByWeightMessageDispatch` works with passed `max_dispatch_weight`.

No, it's not. The `Error` here will early-exit and cause the entire top-level dispatch to fail. I meant to actually take messages from the queue & dispatch them in case the error happens to utilize the `weight` that would otherwise be wasted.

Reply:
The proof is just storage proof && shouldn't fail if some of messages were already delivered. But you're right - I should have utilize `max_dispatch_weigth` even if proof was wrong.

This is irrelevant now.